scarlet/transformations.py

```diff
@@ -15,147 +16,437 @@ def check_cache(name, key):
 
         Parameters
         ----------
-        shape: tuple
-            Shape of the `~scarlet.Source` frame
-        offset_int: int, default=`None`
-            Integer offset of the translation. See `__init__`.
-
+        X: 2D numpy array or `LinearFilter` or `LinearFilterChain`
+            Array to apply the filter to, or chain of filters to
+            prepend this filter to.
+        
         Returns
         -------
-        result: tuple
-            tx, tx_plus, tx_minus, ty, ty_plus, ty_minus, P
-        tx, ty: `~scipy.sparse` array
-            Sparse arrays that contain the diagonal components of the Tx, Ty matrices
-        tx_plus, ty_plus: `~scipy.sparse` array
-            Sparse arrays that contain the off diagonal components of the Tx, Ty matrices.
+        result: 2D numpy array or `LinearFilterChain`
+            If `X` is an array, this is the result of the
+            filter applied to `X`.
+            If `X` is not an image but is another filter
+            (or chain of filters) then a new `LinearFilterChain` is
+            returned with this one prepended.
+        """
+        if isinstance(X, LinearFilter):
+            return LinearFilterChain([self, X])
+        elif isinstance(X, LinearFilterChain):
+            X.filters.insert(0,self)
+            return X
+        else:
+            return self.apply_filter(X)
+
+    def apply_filter(self, X):
+        """Apply the filter to an image
 
+        Parameters
+        ----------
+        X: 2D numpy array or `LinearFilter` or `LinearFilterChain`
+            Array to apply the filter to, or chain of filters to
+            prepend this filter to.
         """
-        self.B, height, width = shape
-        tx = scipy.sparse.diags([1.], offsets=[offset_int[1]], shape=(width, width))
-        tx_minus = scipy.sparse.diags([-1.,1.], offsets=[offset_int[1],offset_int[1]+1], shape=(width, width))
-        tx_plus = scipy.sparse.diags([1.,-1.],offsets=[offset_int[1],offset_int[1]-1], shape=(width, width))
-        tx = scipy.sparse.block_diag([tx]*height)
-        tx_plus = scipy.sparse.block_diag([tx_plus]*height)
-        tx_minus = scipy.sparse.block_diag([tx_minus]*height)
-
-        size = height*width
-        ty = scipy.sparse.diags([1], offsets=[offset_int[0]*width], shape=(size, size), dtype=np.float64)
-        ty_minus = scipy.sparse.diags([-1., 1.], offsets=[offset_int[0]*width, (offset_int[0]+1)*width],
-                                      shape=(size, size))
-        ty_plus = scipy.sparse.diags([1., -1.], offsets=[offset_int[0]*width, (offset_int[0]-1)*width],
-                                     shape=(size, size))
-
-        P = self._adapt_PSF(shape[1:])
-        return tx,tx_plus,tx_minus,ty,ty_plus,ty_minus,P
-
-    def __call__(self, pos, shape, offset_int=None):
-        """Get the operators to translate source
+        return apply_filter(X, self.img.reshape(-1), self.slices, self.inv_slices)
```

Comment:
I find this entire construct rather indirect: `dot` calls `apply_filter` which then calls yet another `apply_filter`. Do you have reasons for this modularization?

Reply:
I would simple code the entire actual operation in `dot`

Reply:
That is what I did originally but I made it modular for 2 different reasons.

1. The `apply_filter` function call was put outside the class in case we find a way to optimize it in C++. 
2. The `apply_filter` is a method in the class in case a particular filter wants to override this method. For example, when I was calculating the runtime using bitmasks and alternative versions specific to the translation operators I wanted to override `apply_filter` but not `dot`, because I did not want to have to reimplement the logic in `dot` that decides how to apply the filter to other filters vs a matrix.

So I'd like to leave it as is to make the framework more flexible.

Reply:
I can see the reasoning behind argument 1, but not for 2. I'd say since these calls are time-critical, I'd streamline them as much as possible. Three function calls to achieve one thing isn't necessary.
Plus: See my comment above about the `reshape(-1)`.